code/iaas/agent-instance/src/main/java/io/cattle/platform/agent/instance/dao/impl/AgentInstanceDaoImpl.java

```diff
@@ -89,19 +101,51 @@ public boolean areAllCredentialsActive(Agent agent) {
 
     @Override
     public List<Long> getAgentProvider(String providedServiceLabel, long accountId) {
-        return Arrays.asList(create().select(INSTANCE.AGENT_ID)
+        final Map<Long, Integer> result = new HashMap<>();
+        create().select(INSTANCE.AGENT_ID, LABEL.VALUE)
                 .from(INSTANCE)
                 .join(INSTANCE_LABEL_MAP)
                     .on(INSTANCE_LABEL_MAP.INSTANCE_ID.eq(INSTANCE.ID))
                 .join(LABEL)
                     .on(LABEL.ID.eq(INSTANCE_LABEL_MAP.LABEL_ID).and(LABEL.KEY.eq(providedServiceLabel)))
                 .where(INSTANCE.ACCOUNT_ID.eq(accountId)
-                    .and(INSTANCE.AGENT_ID.isNotNull())
+                        .and(INSTANCE.AGENT_ID.isNotNull())
                         .and(INSTANCE.STATE.eq(InstanceConstants.STATE_RUNNING))
                         .and(INSTANCE.HEALTH_STATE.in(HealthcheckConstants.HEALTH_STATE_HEALTHY,
                                 HealthcheckConstants.HEALTH_STATE_UPDATING_HEALTHY)))
                 .orderBy(INSTANCE.AGENT_ID.asc())
-                .fetch().intoArray(INSTANCE.AGENT_ID));
+        .fetchInto(new RecordHandler<Record2<Long, String>>() {
+            @Override
+            public void next(Record2<Long, String> record) {
+                Long agentId = record.getValue(INSTANCE.AGENT_ID);
+                String p = record.getValue(LABEL.VALUE);
+                Integer pVal = 1;
+                try {
+                    pVal = Integer.valueOf(p);
+                } catch (Exception e) {
+                    log.warn("Priority value not found for agent {}. Set it as default", agentId);
+                }
+                result.put(agentId, pVal);
+            }
+        });
+        
+        List<Long> returnAgentIds = new ArrayList<>(result.keySet());
+        
+        Collections.sort(returnAgentIds, new Comparator<Long>() {
+            @Override
+            public int compare(Long id1, Long id2) {
+                if (result.get(id1) > result.get(id2)) {
```

Comment:
Instead of re-implementing Integer comparison, we could make use of this method:
https://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html#compareTo(java.lang.Integer)